pyteal/ast/router.py

```diff
@@ -0,0 +1,471 @@
+                        METHOD_ARG_NUM_LIMIT,
+                    )
+                ),
+            )
+            if method_to_register is not None
+            else Txn.application_args.length() == Int(0)
+        )
+
+        # Check if it is a *CREATION*
+        approval_conds: list[Expr] = (
+            [Txn.application_id() == Int(0)] if creation else []
+        )
+        clear_state_conds: list[Expr] = []
+
+        approval_conds.append(method_or_bare_condition)
+
+        # if OC.ClearState exists, add method-or-bare-condition since it is only needed in ClearStateProgram
+        if clear_state_exist:
+            clear_state_conds.append(method_or_bare_condition)
+
+        # Check onComplete conditions for approval_conds, filter out *ClearState*
+        approval_oc_conds: list[Expr] = [
+            Txn.on_completion() == oc
+            for oc in on_completes
+            if str(oc) != str(OnComplete.ClearState)
+        ]
+
+        # if approval OC condition is not empty, append Or to approval_conds
+        if len(approval_oc_conds) > 0:
+            approval_conds.append(Or(*approval_oc_conds))
+
+        # what we have here is:
+        # list of conds for approval program on one branch: creation?, method/bare, Or[OCs]
+        # list of conds for clearState program on one branch: method/bare
+        return approval_conds, clear_state_conds
+
+    @staticmethod
+    def wrap_handler(
+        is_method_call: bool, handler: ABIReturnSubroutine | SubroutineFnWrapper | Expr
+    ) -> Expr:
+        """This is a helper function that handles transaction arguments passing in bare-appcall/abi-method handlers.
+
+        If `is_abi_method` is True, then it can only be `ABIReturnSubroutine`,
+        otherwise:
+            - both `ABIReturnSubroutine` and `Subroutine` takes 0 argument on the stack.
+            - all three cases have none (or void) type.
+
+        On ABI method case, if the ABI method has more than 15 args, this function manages to detuple
+        the last (16-th) Txn app-arg into a list of ABI method arguments, and pass in to the the ABI method.
+
+        Args:
+            is_method_call: a boolean value that specify if the handler is an ABI method.
+            handler: an `ABIReturnSubroutine`, or `SubroutineFnWrapper` (for `Subroutine` case), or an `Expr`.
+        Returns:
+            Expr:
+                - for bare-appcall it returns an expression that the handler takes no txn arg and Approve
+                - for abi-method it returns the txn args correctly decomposed into ABI variables,
+                  passed in ABIReturnSubroutine and logged, then approve.
+        """
+        if not is_method_call:
+            match handler:
+                case Expr():
+                    if handler.type_of() != TealType.none:
+                        raise TealInputError(
+                            "bare appcall handler should have type to be none."
```

Comment:
Add same log as subroutine for error msg?

Reply:
dont understand, since both `Expr` and subroutine case are handled to disallow return of non-`TealType.none`, could you explain more?

Reply:
Just the extra details about what it actually returns besides tealtype.none

Reply:
got it, makes sense